### PR TITLE
feat: optimizer rule to pass expected output ordering hint

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add an optimizer rule `OrderHintRule` that extract expected output ordering from `LogicalPlan::Sort` and passes it to `TableScan` node

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/query/src/optimizer/order_hint.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                    .as_any()
+                    .downcast_ref::<DefaultTableSource>()
+                {
+                    if let Some(adapter) = source
+                        .table_provider
+                        .as_any()
+                        .downcast_ref::<DfTableProviderAdapter>()
+                    {
+                        let schema = adapter.schema();
+                        let mut opts = vec![];
```
Comment (killme2008) [inline src/query/src/optimizer/order_hint.rs:74]:
```suggestion
                        let mut opts = Vec::with_capacity(order_expr.len());
```

File: src/query/src/optimizer/order_hint.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    use datafusion_expr::{col, LogicalPlanBuilder};
+    use datafusion_optimizer::OptimizerContext;
+    use table::table::numbers::NumbersTable;
+
+    use super::*;
+
+    #[test]
+    fn set_order_hint() {
+        let numbers_table = Arc::new(NumbersTable::new(0)) as _;
```
Comment (killme2008) [inline src/query/src/optimizer/order_hint.rs:138]: The test case looks too simple.

File: src/common/recordbatch/src/lib.rs [hunk 3]
```diff
@@ -196,12 +196,17 @@ impl Stream for SimpleRecordBatchStream {
 pub struct RecordBatchStreamAdaptor {
     pub schema: SchemaRef,
     pub stream: Pin<Box<dyn Stream<Item = Result<RecordBatch>> + Send>>,
+    pub output_ordering: Option<Vec<OrderOption>>,
```
Comment (fengjiachun) [inline src/common/recordbatch/src/lib.rs:199]: Why do we need to wrap an `Option` on top of `Vec`? And what does `Some(Empty vec)` mean？

File: src/query/src/optimizer/type_conversion.rs [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+
+/// TypeConversionRule converts some literal values in logical plan to other types according
+/// to data type of corresponding columns.
+/// Specifically:
+/// - string literal of timestamp is converted to `Expr::Literal(ScalarValue::TimestampMillis)`
+/// - string literal of boolean is converted to `Expr::Literal(ScalarValue::Boolean)`
+pub struct TypeConversionRule;
+
+impl AnalyzerRule for TypeConversionRule {
+    fn name(&self) -> &str {
```
Comment (fengjiachun) [inline src/query/src/optimizer/type_conversion.rs:36]: The order of the members does not match that of `AnalyzerRule`.

File: src/common/recordbatch/src/lib.rs (hunk 3 above)
Comment (waynexia) [reply src/common/recordbatch/src/lib.rs:199]: This comes from https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#tymethod.output_ordering. `Some(Empty Vec)` should be the same with `None`

**Review by killme2008 — APPROVED** [review]
LGTM. Still has some clippy warnings.

File: src/table/src/table/adapter.rs [hunk 5]
```diff
@@ -96,8 +101,9 @@ impl TableProvider for DfTableProviderAdapter {
             order_opts
                 .iter()
                 .map(|order_opt| {
-                    let col_name = schema.column_name_by_index(order_opt.index);
-                    let col_expr = Arc::new(Column::new(col_name, order_opt.index));
+                    // let col_name = schema.column_name_by_index(order_opt.index);
```
Comment (v0y4g3r) [inline src/table/src/table/adapter.rs:104]: Can we remove this comment?

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Review by v0y4g3r — APPROVED** [review]

**Pull request merged**